### PR TITLE
Improve ROS2ProjectTemplate

### PR DESCRIPTION
**Pull request by michalpelka**
Improve ROS2ProjectTemplate

- add EditorModule and EditorSystemComponent
- add Sample component that shows integration with ROS2, O3DE, and imgui debugging.

The sample component allows creating in O3DE sample goals for the robot's navigation stack. The goals can be sent via imgui in game mode. It shows integration with basic buses as EntityBus and TransformBus and how to write and publish ROS2 messages from O3DE components' code.
Game mode GUI:
Screenshot from 2023-05-10 11-17-23
Editor:
Screenshot from 2023-05-10 11-20-06

**Note**
- it should be addressed in some doc that HOME key needs to be pressed to get ImGUI.
- It **really** needs a smoke test before merging, since a lot of template code was added.

### CODE REVIEW

**Comment by michalpelka** [comment]
Fixes #246

File: Templates/Ros2ProjectTemplate/Template/Levels/DemoLevel/DemoLevel.prefab [hunk 1]
```diff
@@ -167,6 +168,187 @@
                 }
             }
```
Comment (pawelbudziszewski) [inline Templates/Ros2ProjectTemplate/Template/Levels/DemoLevel/DemoLevel.prefab:169]: It works really great. I have just one suggestion: I'd add marker points (in addition to labels). Something like in the image below. Suggested solution in next comments.

Screenshot from 2023-05-12 21-20-10

File: Templates/Ros2ProjectTemplate/Template/Levels/DemoLevel/DemoLevel.prefab [hunk 2]
```diff
@@ -167,6 +168,187 @@
...
+                    "$type": "EditorVisibilityComponent",
+                    "Id": 5284476042649108894
+                },
+                "Component_[5792898783935716826]": {
+                    "$type": "EditorDisabledCompositionComponent",
+                    "Id": 5792898783935716826
+                },
+                "Component_[9226016495091865703]": {
+                    "$type": "EditorOnlyEntityComponent",
+                    "Id": 9226016495091865703
```
Comment (pawelbudziszewski) [inline Templates/Ros2ProjectTemplate/Template/Levels/DemoLevel/DemoLevel.prefab:289]:
```suggestion
                    "Id": 3950088389572830633,
                    "Child Entity Order": [
                        "Entity_[558067385513]"
                    ]
```

File: Templates/Ros2ProjectTemplate/Template/Levels/DemoLevel/DemoLevel.prefab [hunk 3]
```diff
@@ -167,6 +168,187 @@
                 }
             }
         },
```
Comment (pawelbudziszewski) [inline Templates/Ros2ProjectTemplate/Template/Levels/DemoLevel/DemoLevel.prefab:170]:
```suggestion
        },
        "Entity_[519412679849]": {
            "Id": "Entity_[519412679849]",
            "Name": "marker",
            "Components": {
                "Component_[12345039422781644436]": {
                    "$type": "{27F1E1A1-8D9D-4C3B-BD3A-AFB9762449C0} TransformComponent",
                    "Id": 12345039422781644436,
                    "Parent Entity": "Entity_[549827034797]",
                    "Transform Data": {
                        "UniformScale": 0.10000000149011612
                    }
                },
                "Component_[16311924938308571969]": {
                    "$type": "AZ::Render::EditorMeshComponent",
                    "Id": 16311924938308571969,
                    "Controller": {
                        "Configuration": {
                            "ModelAsset": {
                                "assetId": {
                                    "guid": "{1EE46427-D1C2-5698-B30D-B4CCF46B15CD}",
                                    "subId": 274434668
                                },
                                "assetHint": "objects/_primitives/_sphere_1x1.azmodel"
                            }
                        }
                    }
                },
                "Component_[17981273470482180971]": {
                    "$type": "EditorEntitySortComponent",
                    "Id": 17981273470482180971
                },
                "Component_[2065222331617120332]": {
                    "$type": "EditorVisibilityComponent",
                    "Id": 2065222331617120332
                },
                "Component_[2096811794606887530]": {
                    "$type": "EditorDisabledCompositionComponent",
                    "Id": 2096811794606887530
                },
                "Component_[2855690690078728486]": {
                    "$type": "EditorPendingCompositionComponent",
                    "Id": 2855690690078728486
                },
                "Component_[2949627029165663181]": {
                    "$type": "EditorMaterialComponent",
                    "Id": 2949627029165663181,
                    "Controller": {
                        "Configuration": {
                            "materials": [
                                {
                                    "Key": {
                                        "materialSlotStableId": 1738373820
                                    },
                                    "Value": {
                                        "MaterialAsset": {
                                            "assetId": {
                                                "guid": "{9C47066E-BD8F-5C1B-B935-933296BBE312}"
                                            },
                                            "assetHint": "materials/presets/macbeth/15_red.azmaterial"
                                        }
                                    }
                                }
                            ]
                        }
                    }
                },
                "Component_[3984524579386645338]": {
                    "$type": "EditorLockComponent",
                    "Id": 3984524579386645338
                },
                "Component_[6180558729729631244]": {
                    "$type": "EditorInspectorComponent",
                    "Id": 6180558729729631244
                },
                "Component_[6203285465432256414]": {
                    "$type": "EditorOnlyEntityComponent",
                    "Id": 6203285465432256414
                },
                "Component_[7639289353853227968]": {
                    "$type": "EditorEntityIconComponent",
                    "Id": 7639289353853227968
                }
            }
        },
        "Entity_[558067385513]": {
            "Id": "Entity_[558067385513]",
            "Name": "marker",
            "Components": {
                "Component_[12345039422781644436]": {
                    "$type": "{27F1E1A1-8D9D-4C3B-BD3A-AFB9762449C0} TransformComponent",
                    "Id": 12345039422781644436,
                    "Parent Entity": "Entity_[545532067501]",
                    "Transform Data": {
                        "UniformScale": 0.10000000149011612
                    }
                },
                "Component_[16311924938308571969]": {
                    "$type": "AZ::Render::EditorMeshComponent",
                    "Id": 16311924938308571969,
                    "Controller": {
                        "Configuration": {
                            "ModelAsset": {
                                "assetId": {
                                    "guid": "{1EE46427-D1C2-5698-B30D-B4CCF46B15CD}",
                                    "subId": 274434668
                                },
                                "assetHint": "objects/_primitives/_sphere_1x1.azmodel"
                            }
                        }
                    }
                },
                "Component_[17981273470482180971]": {
                    "$type": "EditorEntitySortComponent",
                    "Id": 17981273470482180971
                },
                "Component_[2065222331617120332]": {
                    "$type": "EditorVisibilityComponent",
                    "Id": 2065222331617120332
                },
                "Component_[2096811794606887530]": {
                    "$type": "EditorDisabledCompositionComponent",
                    "Id": 2096811794606887530
                },
                "Component_[2855690690078728486]": {
                    "$type": "EditorPendingCompositionComponent",
                    "Id": 2855690690078728486
                },
                "Component_[2915573670259884143]": {
                    "$type": "EditorMaterialComponent",
                    "Id": 2915573670259884143,
                    "Controller": {
                        "Configuration": {
                            "materials": [
                                {
                                    "Key": {
                                        "materialSlotStableId": 1738373820
                                    },
                                    "Value": {
                                        "MaterialAsset": {
                                            "assetId": {
                                                "guid": "{9C47066E-BD8F-5C1B-B935-933296BBE312}"
                                            }
                                        }
                                    }
                                }
                            ]
                        }
                    }
                },
                "Component_[3984524579386645338]": {
                    "$type": "EditorLockComponent",
                    "Id": 3984524579386645338
                },
                "Component_[6180558729729631244]": {
                    "$type": "EditorInspectorComponent",
                    "Id": 6180558729729631244
                },
                "Component_[6203285465432256414]": {
                    "$type": "EditorOnlyEntityComponent",
                    "Id": 6203285465432256414
                },
                "Component_[7639289353853227968]": {
                    "$type": "EditorEntityIconComponent",
                    "Id": 7639289353853227968
                }
            }
        },                
```

File: Templates/Ros2ProjectTemplate/Template/Levels/DemoLevel/DemoLevel.prefab [hunk 4]
```diff
@@ -167,6 +168,187 @@
...
+                    "$type": "EditorOnlyEntityComponent",
+                    "Id": 16607348232201416003
+                },
+                "Component_[16857701992694867242]": {
+                    "$type": "EditorInspectorComponent",
+                    "Id": 16857701992694867242
+                },
+                "Component_[18080104825946547885]": {
+                    "$type": "EditorEntitySortComponent",
+                    "Id": 18080104825946547885
```
Comment (pawelbudziszewski) [inline Templates/Ros2ProjectTemplate/Template/Levels/DemoLevel/DemoLevel.prefab:319]:
```suggestion
                    "Id": 18080104825946547885,
                    "Child Entity Order": [
                        "Entity_[519412679849]"
                    ]
```

File: Templates/Ros2ProjectTemplate/Template/Levels/DemoLevel/DemoLevel.prefab (hunk 1 above)
Comment (adamdbrw) [reply Templates/Ros2ProjectTemplate/Template/Levels/DemoLevel/DemoLevel.prefab:169]: Agreed

**Review by pawelbudziszewski — APPROVED** [review]

**Comment by michalpelka** [comment]
Smoke tested Pawel's contribution.

File: Templates/Ros2ProjectTemplate/Template/Gem/Source/${Name}SampleComponent.cpp [hunk 5]
```diff
@@ -0,0 +1,86 @@
+
+
+#include "${SanitizedCppName}SampleComponent.h"
+#include <AzCore/Component/ComponentApplicationBus.h>
+#include <AzCore/Serialization/EditContext.h>
+#include <AzCore/Serialization/SerializeContext.h>
+
+#include <AzCore/Component/TransformBus.h>
+#include <AzCore/Math/Transform.h>
```
Comment (spham-amzn) [inline Templates/Ros2ProjectTemplate/Template/Gem/Source/${Name}SampleComponent.cpp:9]: Lets group these includes together alphabetically
```suggestion
#include <AzCore/Component/ComponentApplicationBus.h>
#include <AzCore/Component/TransformBus.h>
#include <AzCore/Math/Transform.h>
#include <AzCore/Serialization/EditContext.h>
#include <AzCore/Serialization/SerializeContext.h>
```

File: Templates/Ros2ProjectTemplate/Template/Gem/Source/${Name}SampleComponent.cpp [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+namespace ${SanitizedCppName}
+{
+
+    ${SanitizedCppName}SampleComponent::${SanitizedCppName}SampleComponent()
+    {
+        m_goalTopicConfiguration.m_topic = "/goal_pose";
+        m_goalTopicConfiguration.m_type = "geometry_msgs::msg::PoseStamped";
+    }
+    void ${SanitizedCppName}SampleComponent::Activate()
```
Comment (spham-amzn) [inline Templates/Ros2ProjectTemplate/Template/Gem/Source/${Name}SampleComponent.cpp:26]: nit: Add a newline between each function

File: Templates/Ros2ProjectTemplate/Template/Gem/Source/${Name}SampleComponent.cpp [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+                ec->Class<${SanitizedCppName}SampleComponent>("${SanitizedCppName}SampleComponent", "A sample component that sends goal")
+                    ->ClassElement(AZ::Edit::ClassElements::EditorData, "${SanitizedCppName}SampleComponent")
+                    ->Attribute(AZ::Edit::Attributes::AppearsInAddComponentMenu, AZ_CRC("Game"))
+                    ->Attribute(AZ::Edit::Attributes::Category, "Sample ROS2 Project")
+                    ->DataElement(AZ::Edit::UIHandlers::Default, &${SanitizedCppName}SampleComponent::m_goalEntities, "Goals", "Entities goals")
+                    ->DataElement(
+                        AZ::Edit::UIHandlers::Default,
+                        &${SanitizedCppName}SampleComponent::m_goalTopicConfiguration,
+                        "Topic for goal message",
+                        "Configuration for ros2 topic to send goal message to");
```
Comment (spham-amzn) [inline Templates/Ros2ProjectTemplate/Template/Gem/Source/${Name}SampleComponent.cpp:60]: `ros2` should be `ROS2` for consistency right?

File: Templates/Ros2ProjectTemplate/Template/Gem/Source/${Name}SampleComponent.h [hunk 8]
```diff
@@ -0,0 +1,40 @@
+
```
Comment (spham-amzn) [inline Templates/Ros2ProjectTemplate/Template/Gem/Source/${Name}SampleComponent.h:1]: Missing copyright header
```suggestion

// {BEGIN_LICENSE}
/*
 * Copyright (c) Contributors to the Open 3D Engine Project.
 * For complete copyright and license terms please see the LICENSE at the root of this distribution.
 *
 * SPDX-License-Identifier: Apache-2.0 OR MIT
 *
 */
// {END_LICENSE}
```

**Review by adamdbrw — APPROVED** [review]

**Pull request merged**